Default to a 0 evaluation frequency in dwh alerts

moolitayer opened the pull request:
See https://github.com/ManageIQ/manageiq/pull/14318 for explanation why a 0 notification frequency makes sense here

![manageiq control](https://cloud.githubusercontent.com/assets/3010449/24139830/59538eca-0e26-11e7-86ef-b7d25c5fa5cc.png)

moolitayer commented:
@miq-bot add_label bug, providers/containers

moolitayer commented:
@simon3z please review, this has no dependencies

moolitayer commented on app/controllers/miq_policy_controller/alerts.rb:
```diff
@@ -486,7 +491,7 @@ def alert_build_pulldowns
       15.minutes.to_i => _("15 Minutes"), 30.minutes.to_i => _("30 Minutes"), 1.hour.to_i => _("1 Hour"),
       2.hours.to_i => _("2 Hours"), 3.hours.to_i => _("3 Hours"), 4.hours.to_i => _("4 Hours"),
-      6.hours.to_i => _("6 Hours"), 12.hours.to_i => _("12 Hours"), 1.day.to_i => _("1 Day")
+      6.hours.to_i => _("6 Hours"), 12.hours.to_i => _("12 Hours"), 1.day.to_i => _("1 Day"), 0.minutes.to_i => _("0")
```
@gtanzillo is it ok to add a 0 notification frequency for existing alerts types? 

also note it won't be the default

moolitayer commented:
@cben @zgalor please review

simon3z commented:
@moolitayer does it deserve a better naming in the UI rather than "0"? Can we think of something more descriptive? (I don't know maybe "Immediate"?)
What about other alerts? Is this option available only for DWH?

moolitayer commented:
@simon3z I'm not sure what's most usable here. I added a screen shot, the values are
`Notification Frequency: 0, Notification Frequency: 1 Minute ...`

Another option I can think of is `Notification Frequency: Always Evaluate`

simon3z commented:
@moolitayer what about other alerts? Is this option available only for DWH?

@moolitayer for DWH what are the options you display? If "0" is the only one that makes sense, is it the only one available?

moolitayer commented:
> @moolitayer what about other alerts? Is this option available only for DWH?
> 

I [asked](https://github.com/ManageIQ/manageiq-ui-classic/pull/678#discussion_r106928615) about that but I guess it is better to only have this options for dwh. Updated the PR.

> @moolitayer for DWH what are the options you display? If "0" is the only one that makes sense, is it the only one available?

Updated. Note since our main use case is API I initial created https://github.com/ManageIQ/manageiq/pull/14318 to block non zero alerts in the model

simon3z commented:
@moolitayer are all these validations enforced for Rest-API as well? (I suppose they should be at model-level for that)

moolitayer commented:
@simon3z since our main use case is API I initial created ManageIQ/manageiq#14318 to block non zero alerts in the model

simon3z commented:
@moolitayer not sure what "Always Notify" means, could it be better saying "Notify Always"?
Also, Always/In some cases/Never is not an interval frequency so it doesn't seem to belong there.

moolitayer commented:
> Also, Always/In some cases/Never is not an interval frequency so it doesn't seem to belong there

So back to 0 then?

simon3z commented:
> > Also, Always/In some cases/Never is not an interval frequency so it doesn't seem to belong there
> 
> So back to 0 then?

@moolitayer I wanted to come up with a better naming... what is this doing? This is evaluating the alerts as soon as received... so maybe something along the lines of "Upon Receipt", "When Received", "Immediate", etc. Maybe someone from UX can help us (provided that we explain what this is about).

moolitayer commented:
Notification frequency does not describe what this value means unless i'm misreading it.

If an alert's generating event happened and it has not yet been this time since last evaluation[1], it is not evaluated (MiqAlertStatus will not be created, actions will not fire, etc). see https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_alert.rb#L161

Something parallel in ovirt engine's event log is named flood rate.

While this name is what it is the best I can think of is `[Always Notify, Always Evaluate, 0, -1]`

@simon3z cc

[1] for the same target object ofcurse

simon3z commented:
@moolitayer since it seems that it's not really configurable (frequency is externally defined in the hawkular triggers) then we should probably hide the option altogether.

moolitayer commented:
@simon3z nice, done

moolitayer commented:
Rubocop fix

moolitayer commented:
attempting to assign again per [comment](https://github.com/ManageIQ/manageiq-ui-classic/pull/678#pullrequestreview-28425952)

@miq-bot assign @gtanzillo 

@gtanzillo Please take a look

moolitayer commented:
attempting to assign again per [comment](https://github.com/ManageIQ/manageiq-ui-classic/pull/678#pullrequestreview-28425952)

@miq-bot assign gtanzillo 

@gtanzillo Please take a look

gtanzillo commented on app/controllers/miq_policy_controller/alerts.rb:
```diff
@@ -345,9 +345,14 @@ def alert_build_edit_screen
+    if (@edit[:new][:expression][:eval_method] && @edit[:new][:expression][:eval_method] == "hourly_performance") ||
+       @edit[:new][:exp_event] == "_hourly_timer_"
+      1.hour.to_i
+    elsif @edit[:new][:expression][:eval_method] && @edit[:new][:expression][:eval_method] == "dwh_generic"
```
@moolitayer Are you able to check `:always_evaluate` here instead? This way the Ui code can also be generic.

moolitayer replied:
The UI code does not send or knows about :always_evaluate, that is only on the server side

moolitayer replied:
@gtanzillo PTAL

moolitayer replied:
@gtanzillo Please take another look.

gtanzillo replied:
Understood @moolitayer

gtanzillo commented on app/views/miq_policy/_alert_details.html.haml:
```diff
@@ -101,13 +101,17 @@
             .col-md-8
               - if @edit[:new][:expression][:eval_method] == "hourly_performance"
                 - repeat = @sb[:alert][:hourly_repeat_times]
+              - elsif @edit[:new][:expression][:eval_method] == "dwh_generic"
```
Same here - would be better if you can check `:always_evaluate` here.

moolitayer commented:
@dclarizio please reivew

moolitayer commented:
@dclarizio please review
